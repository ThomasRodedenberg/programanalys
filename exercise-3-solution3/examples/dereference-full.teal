type T(x : string, y : int);


fun f(a : nonnull array[int], y : int) : int = {
  if y < 0 {
    return -1;
  }

  return a[y];
}

fun main() = {
  var a : T := new T("", f(null, -1));
  var b : T := new T("", 0);
  var c : nonnull T := b;
  var d : T := new T("", null);

  b := null;
  d := c;
  b := a;

  f([1, 2], a.y);
  f([1, 2], b.y);
  f([1, 2], c.y);
  f([1, 2], d.y);

  var aa := [];
  if 1 {
    aa := null;
  }
  aa[0] := 1;
  var bb := aa[0];

  a.y := 1;
  return b.x;
}
