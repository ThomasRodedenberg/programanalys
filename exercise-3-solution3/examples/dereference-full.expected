merge: 1 7 19
N 4 6 -> null 13 27
N 4 6 -> obj 22 4
N 4 6 -> obj 23 4
N 4 6 -> obj 24 4
N 4 6 -> obj 25 4
maybe: NI 9 9 -> obj 6 11
maybe: NI 9 9 -> obj 13 21
maybe: NI 9 9 -> obj 13 33
maybe: NI 9 9 -> obj 14 21
maybe: NI 9 9 -> obj 14 25
maybe: NI 9 9 -> obj 16 21
NI 9 9 -> null 16 25
maybe: NI 9 9 -> obj 22 5
maybe: NI 9 9 -> obj 22 8
maybe: NI 9 9 -> obj 23 5
maybe: NI 9 9 -> obj 23 8
maybe: NI 9 9 -> obj 24 5
maybe: NI 9 9 -> obj 24 8
maybe: NI 9 9 -> obj 25 5
maybe: NI 9 9 -> obj 25 8
maybe: NI 9 9 -> obj 34 9
maybe: NA 9 9 -> null 13 27
maybe: NA 9 9 -> obj 22 4
maybe: NA 9 9 -> obj 23 4
maybe: NA 9 9 -> obj 24 4
maybe: NA 9 9 -> obj 25 4
N 15 2 -> obj 13 15
N 15 2 -> obj 14 15
N 15 2 -> null 18 7
maybe: NO 23 12 -> obj 13 15
maybe: NO 23 12 -> obj 14 15
maybe: NO 23 12 -> null 18 7
maybe: NO 24 12 -> obj 13 15
maybe: NO 24 12 -> obj 14 15
maybe: NO 24 12 -> null 18 7
maybe: NO 25 12 -> obj 13 15
maybe: NO 25 12 -> obj 14 15
maybe: NO 25 12 -> obj 16 15
NO 25 12 -> null 18 7
maybe: NA 31 2 -> obj 27 12
NA 31 2 -> null 29 10
maybe: NA 32 12 -> obj 27 12
NA 32 12 -> null 29 10
maybe: NO 35 9 -> obj 13 15
maybe: NO 35 9 -> obj 14 15
maybe: NO 35 9 -> null 18 7
